tests/test_types.py

```diff
@@ -5369,61 +5362,103 @@ class Model(BaseModel):
 
     with pytest.raises(ValidationError) as exc_info:
         Model(lower='')
+    # insert_assert(exc_info.value.errors(include_url=False))
     assert exc_info.value.errors(include_url=False) == [
         {
-            'ctx': {'min_length': 1},
-            'input': '',
+            'type': 'too_short',
             'loc': ('lower',),
-            'msg': 'String should have at least 1 characters',
-            'type': 'string_too_short',
+            'msg': 'Value should have at least 1 item after validation, not 0',
+            'input': '',
+            'ctx': {'field_type': 'Value', 'min_length': 1, 'actual_length': 0},
+        }
+    ]
+
+
+def test_constraint_dataclass() -> None:
+    @dataclass(order=True)
+    # need to make it inherit from int so that
+    # because the PydanticKnownError requires it to be a number
+    # but it's not really relevant to this test
+    class MyDataclass(int):
+        x: int
+
+    ta = TypeAdapter(Annotated[MyDataclass, annotated_types.Gt(MyDataclass(0))])
+
+    assert ta.validate_python(MyDataclass(1)) == MyDataclass(1)
+
+    with pytest.raises(ValidationError) as exc_info:
+        ta.validate_python(MyDataclass(0))
+
+    # insert_assert(exc_info.value.errors(include_url=False))
+    assert exc_info.value.errors(include_url=False) == [
+        {
+            'type': 'greater_than',
+            'loc': (),
+            'msg': 'Input should be greater than 0',
+            'input': MyDataclass(0),
+            'ctx': {'gt': MyDataclass(0)},
         }
     ]
 
 
 def test_transform_schema_for_third_party_class():
     # Here, third party means you can't define methods on the class directly, so have to use annotations.
 
-    class DatetimeWrapper:
+    class IntWrapper:
```

Comment:
I had to change from datetime to int because `PydanticKnownError` w/ `gt` only accepts numeric types. We can either use a `PydanticCustomError` here which is mostly API compatible but might take a performance hit (when I first implemented the functions in `_validators.py` @samuelcolvin pushed me to use `PydanticKnownError` over `PydanticCustomError` for this reason) or we can change `PydanticKnownError` in `pydantic-core` to not be as pedantic.